vowpalwabbit/reductions/epsilon_decay.cc

```diff
@@ -0,0 +1,247 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "epsilon_decay.h"
+
+#include "config/options.h"
+#include "global_data.h"
+#include "label_type.h"
+#include "learner.h"
+#include "memory.h"
+#include "model_utils.h"
+#include "prediction_type.h"
+#include "vw.h"
+
+#include <algorithm>
+#include <cmath>
+#include <string>
+
+using namespace VW::config;
+using namespace VW::LEARNER;
+
+// namespace logger = VW::io::logger;
+
+namespace VW
+{
+namespace epsilon_decay
+{
+void epsilon_decay_score::update_bounds(float w, float r)
+{
+  update(w, r);
+
+  // update the lower bound
+  distributionally_robust::ScoredDual sd = this->chisq.recompute_duals();
+  _lower_bound = static_cast<float>(sd.first);
+}
+
+template <class ForwardIt>
+ForwardIt swap_models(ForwardIt first, ForwardIt n_first, ForwardIt end)
+{
+  for (; first != end; ++first, ++n_first) { std::iter_swap(first, n_first); }
+  return n_first;
+}
+
+template <class ForwardIt>
+void reset_models(ForwardIt first, ForwardIt end, parameters& weights, double epsilon_decay_alpha,
+    double epsilon_decay_tau, uint64_t model_count)
+{
+  uint64_t params_per_weight = 1;
+  while (params_per_weight < model_count) { params_per_weight *= 2; }
+  for (; first != end; ++first)
+  {
+    first->reset_stats(epsilon_decay_alpha, epsilon_decay_tau);
+    weights.dense_weights.clear_offset(first->get_model_idx(), params_per_weight);
+  }
+}
+
+float decayed_epsilon(uint64_t update_count) { return static_cast<float>(std::pow(update_count + 1, -1.f / 3.f)); }
```

Comment:
Is there a paper with reference to why 1/cube_root(n) makes sense?  A reference to that here would be useful.

Reply:
I'll ask John in the next sync